### PR TITLE
Add path utils and test config for filesystem

### PR DESCRIPTION
**Pull request by olaservo**
## Description

Separating out the test config and utils from this older PR to simplify the actual changes needed to the server: https://github.com/modelcontextprotocol/servers/pull/543

Plan is to merge this, then the original PR author can update their branch to reference the utils (or create a new one with just those changes).

## Server Details
- Server: filesystem
- Changes to: tests

## Motivation and Context

Original PR includes a large number of changes so the intention is to split it up so we can finally integrate the Windows path fixes into the server.

## How Has This Been Tested?

```
 PASS  __tests__/path-utils.test.ts
  Path Utilities
    convertToWindowsPath
      √ leaves Unix paths unchanged (4 ms)
      √ converts WSL paths to Windows format (1 ms)
      √ converts Unix-style Windows paths to Windows format
      √ leaves Windows paths unchanged but ensures backslashes
      √ handles Windows paths with spaces
      √ handles uppercase and lowercase drive letters
    normalizePath
      √ preserves Unix paths (1 ms)
      √ removes surrounding quotes (1 ms)
      √ normalizes backslashes (1 ms)
      √ converts forward slashes to backslashes on Windows (1 ms)
      √ handles WSL paths
      √ handles Unix-style Windows paths
      √ handles paths with spaces and mixed slashes (1 ms)
      √ preserves spaces in all path formats (1 ms)
      √ handles special characters in paths (1 ms)
    expandHome
      √ expands ~ to home directory (1 ms)
      √ leaves other paths unchanged (1 ms)

Test Suites: 1 passed, 1 total
Tests:       17 passed, 17 total
Snapshots:   0 total
Time:        6.597 s
Ran all test suites.
```

## Breaking Changes
None

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
- [x] I have read the [MCP Protocol Documentation](https://modelcontextprotocol.io)
- [x] My changes follows MCP security best practices
- [x] I have updated the server's README accordingly
- [ ] I have tested this with an LLM client
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [ ] I have documented all environment variables and configuration options

### CODE REVIEW

**Comment by cliffhall** [comment]
The tests all pass locally on macOS. A few observations/questions.

* Is `file-operations.test.ts` just meant to test the node.js functionality on both platforms? 
* Is `core-functionality.ts` meant to test the way the server does certain things by replicating the logic inside the tests? 
* I don't see any external project code that `file-operations.test.ts` and `core-functionality.ts`  are testing.
  - For instance in `file-operations.test.ts` there is a test called `formatSize helper function` which tests a function the test itself defines, but isn't located anywhere else in the project.

* I turned on coverage locally just to see where we were, and saw that `path-utils.test.ts` has a couple of untested branches / lines.
Screenshot 2025-06-17 at 1 46 08 PM

* Gemini suggested these tests to cover those: 
```typescript


   // New test to ensure drive letter capitalization (hits line 67 true, and line 70)
    it('capitalizes lowercase drive letters for Windows paths', () => {
      expect(normalizePath('c:/windows/system32'))
          .toBe('C:\\windows\\system32');
      expect(normalizePath('/mnt/d/my/folder')) // WSL path with lowercase drive
          .toBe('D:\\my\\folder');
      expect(normalizePath('/e/another/folder')) // Unix-style Windows path with lowercase drive
          .toBe('E:\\another\\folder');
    });

    // New test for the 'else' branch of the condition on line 67 (hits line 73)
    it('returns normalized non-Windows/WSL/Unix-style Windows paths as is after basic normalization', () => {
      // Relative path
      const relativePath = 'some/relative/path';
      expect(normalizePath(relativePath)).toBe(path.normalize(relativePath));

      // UNC Path (should not match the regex on line 67)
      const uncPath = '\\\\SERVER\\share\\folder';
      // Expect the result of path.normalize on the UNC path
      expect(normalizePath(uncPath)).toBe(path.normalize(uncPath));

      // A path that looks somewhat absolute but isn't a drive or recognized Unix root for Windows conversion
      const otherAbsolutePath = '\\someserver\\share\\file';
      expect(normalizePath(otherAbsolutePath)).toBe(path.normalize(otherAbsolutePath));
    });
```  

  However, that last test failed with: 

```shell
 Path Utilities › normalizePath › returns normalized non-Windows/WSL/Unix-style Windows paths as is after basic normalization

    expect(received).toBe(expected) // Object.is equality

    Expected: "\\\\SERVER\\share\\folder"
    Received: "\\SERVER\\share\\folder"

      149 |       const uncPath = '\\\\SERVER\\share\\folder';
      150 |       // Expect the result of path.normalize on the UNC path
    > 151 |       expect(normalizePath(uncPath)).toBe(path.normalize(uncPath));
          |                                      ^
      152 |
      153 |       // A path that looks somewhat absolute but isn't a drive or recognized Unix root for Windows conversion
      154 |       const otherAbsolutePath = '\\someserver\\share\\file';

      at Object.<anonymous> (__tests__/path-utils.test.ts:151:38)
 ```  
  
When I asked it for a fix, it pointed out a possible flaw 

> `The test failure indicates that your normalizePath function is altering UNC paths (like \\\\SERVER\\share\\folder) in a way that differs from the standard path.normalize() behavior on your system. Specifically, normalizePath seems to be reducing the leading \\\\ to \\.`

> `This is likely due to the line p = p.replace(/\\\\/g, '\\'); in your normalizePath function in src/filesystem/path-utils.ts. This replacement is too general and affects the leading double backslash of UNC paths.` 

> `To fix this and ensure UNC paths are handled correctly (preserving their \\\\ prefix), you should modify the normalizePath function to be more specific when replacing double backslashes.`
  
  
The fix it suggested for `path-utils.ts` was
  
Screenshot 2025-06-17 at 2 02 49 PM

 ```typescript
 // Convert WSL or Unix-style Windows paths to Windows format
  p = convertToWindowsPath(p);

  // Handle double backslashes, preserving leading UNC \\
  if (p.startsWith('\\\\')) {
    // For UNC paths, normalize double backslashes *after* the leading marker
    const restOfPath = p.substring(2).replace(/\\\\/g, '\\');
    p = '\\\\' + restOfPath;
  } else {
    // For non-UNC paths, normalize all double backslashes
    p = p.replace(/\\\\/g, '\\');
}
```


With those tests added and the fix above in place, 100% coverage on `path-utils.ts`
Screenshot 2025-06-17 at 2 05 21 PM

**Comment by olaservo** [comment]
> The tests all pass locally on macOS. A few observations/questions.
> 
> * Is `file-operations.test.ts` just meant to test the node.js functionality on both platforms?
> * Is `core-functionality.ts` meant to test the way the server does certain things by replicating the logic inside the tests?
> * I don't see any external project code that `file-operations.test.ts` and `core-functionality.ts`  are testing.
>   
>   * For instance in `file-operations.test.ts` there is a test called `formatSize helper function` which tests a function the test itself defines, but isn't located anywhere else in the project.
> * I turned on coverage locally just to see where we were, and saw that `path-utils.test.ts` has a couple of untested branches / lines.


Thanks for checking this out, I originally created the PR to add tests for another PR that someone else had opened related to better Windows handling: https://github.com/modelcontextprotocol/servers/pull/543

I think things got weird with the state of the PR and I obviously didn't look closely enough at what part of this was testing, I appreciate that you did.  I think its all good now?

**Comment by cliffhall** [comment]
Those changes were a bit different from the suggested ones that passed locally.

I'm getting these errors now:

```
 FAIL  __tests__/path-utils.test.ts
  Path Utilities
    convertToWindowsPath
      ✓ leaves Unix paths unchanged (1 ms)
      ✓ converts WSL paths to Windows format
      ✓ converts Unix-style Windows paths to Windows format
      ✓ leaves Windows paths unchanged but ensures backslashes
      ✓ handles Windows paths with spaces
      ✓ handles uppercase and lowercase drive letters
    normalizePath
      ✓ preserves Unix paths
      ✓ removes surrounding quotes
      ✓ normalizes backslashes
      ✓ converts forward slashes to backslashes on Windows
      ✓ handles WSL paths
      ✓ handles Unix-style Windows paths (1 ms)
      ✓ handles paths with spaces and mixed slashes
      ✓ preserves spaces in all path formats
      ✓ handles special characters in paths
      ✓ capitalizes lowercase drive letters for Windows paths (1 ms)
      ✕ handles UNC paths correctly (1 ms)
      ✕ returns normalized non-Windows/WSL/Unix-style Windows paths as is after basic normalization (1 ms)
    expandHome
      ✓ expands ~ to home directory
      ✓ leaves other paths unchanged

  ● Path Utilities › normalizePath › handles UNC paths correctly

    expect(received).toBe(expected) // Object.is equality

    Expected: "\\\\SERVER\\share\\folder"
    Received: "\\\\\\SERVER\\share\\folder"

      142 |       // Test UNC path with double backslashes that need normalization
      143 |       const uncPathWithDoubles = '\\\\\\\\SERVER\\\\share\\\\folder';
    > 144 |       expect(normalizePath(uncPathWithDoubles)).toBe('\\\\SERVER\\share\\folder');
          |                                                 ^
      145 |     });
      146 |
      147 |     it('returns normalized non-Windows/WSL/Unix-style Windows paths as is after basic normalization', () => {

      at Object.<anonymous> (__tests__/path-utils.test.ts:144:49)

  ● Path Utilities › normalizePath › returns normalized non-Windows/WSL/Unix-style Windows paths as is after basic normalization

    expect(received).toBe(expected) // Object.is equality

    Expected: "some\\relative\\path"
    Received: "some/relative/path"

      148 |       // Relative path
      149 |       const relativePath = 'some/relative/path';
    > 150 |       expect(normalizePath(relativePath)).toBe(relativePath.replace(/\//g, '\\'));
          |                                           ^
      151 |
      152 |       // A path that looks somewhat absolute but isn't a drive or recognized Unix root for Windows conversion
      153 |       const otherAbsolutePath = '\\someserver\\share\\file';

      at Object.<anonymous> (__tests__/path-utils.test.ts:150:43)
```
Analysis of the current failures:

1. **Test:** `handles UNC paths correctly`:
* **Input:** `\\\\\\\\SERVER\\\\share\\\\folder` (excessive leading and internal backslashes)
* **Expected:** `\\\\SERVER\\share\\folder`
* **Received:** \\\\\\SERVER\\share\\folder (one too many leading backslashes)
* The current UNC handling logic in normalizePath seems to be miscounting or incorrectly re-adding a leading backslash when `path.normalize` interacts with multiple initial backslashes, or the subsequent regex replacement isn't quite right for all edge cases of `path.normalize`'s output.

2. **Test:** `returns normalized non-Windows/WSL/Unix-style Windows paths as is after basic normalization` (for relative paths):
* Input: `some/relative/path`
* **Expected:** `some\\relative\\path` (Windows-style, as per the test's expectation)
* **Received:** `some/relative/path`
* The function currently returns `path.normalize('some/relative/path')`. On a Unix-like system (or a Node environment where path.normalize preserves forward slashes for relative paths), this will be `some/relative/path`. The final block that converts / to \ only triggers if the path already looks like a Windows absolute path (e.g., starts with a drive letter). The test implies that if a path is not an explicit "Unix path" (as per isUnixPath), it should be treated as a Windows path, including converting slashes in relative paths.

I'll add the suggestions to `path-utils.ts` that make the tests pass.

File: src/filesystem/path-utils.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    // Capitalize drive letter if present
+    if (/^[a-z]:/.test(result)) {
+      result = result.charAt(0).toUpperCase() + result.slice(1);
+    }
+    return result;
+  }
+  
+  // Leave other paths unchanged
+  return normalized;
+}
```
Comment (cliffhall) [inline src/filesystem/path-utils.ts]:
```suggestion
export function normalizePath(p: string): string {
  // Remove any surrounding quotes and whitespace
  p = p.trim().replace(/^["']|["']$/g, '');

  // Check if this is a "pure" Unix path (starts with / but not a Windows-emulated path like /c/ or /mnt/c/)
  const isPureUnixPath = p.startsWith('/') &&
      !p.match(/^\/mnt\/[a-z]\//i) &&
      !p.match(/^\/[a-zA-Z]\//);

  if (isPureUnixPath) {
    // For pure Unix paths, normalize multiple forward slashes and remove trailing ones (unless it's the root).
    // Keep forward slashes.
    let result = p.replace(/\/+/g, '/');
    if (result !== '/' && result.endsWith('/')) {
      result = result.slice(0, -1);
    }
    return result;
  }

  // For Windows-style paths or relative paths that should use Windows conventions.
  let tempPath = convertToWindowsPath(p); // Handles /mnt/c, /c/, C:/ -> C:\
  const wasInputUNC = tempPath.startsWith('\\\\');

  // Before path.normalize, clean up excessive internal backslashes for non-UNC paths.
  // For UNC paths, path.normalize is expected to handle the prefix variations.
  if (!wasInputUNC) {
    tempPath = tempPath.replace(/\\\\/g, '\\');
  }

  let normalized = path.normalize(tempPath);

  if (wasInputUNC) {
    // Ensure UNC path starts with exactly \\ and has single \ internally.
    // path.normalize might return \\server\share, \\?\UNC\server\share, or even \server\share from an input like \\\\input.
    let serverAndShare = normalized.replace(/^\\+(\?\UNC\\)?/, ''); // Remove leading \ and optional ?\UNC\ prefix
    serverAndShare = serverAndShare.replace(/[\\\/]+/g, '\\');    // Normalize all internal separators to single \
    normalized = '\\\\' + serverAndShare;
  } else {
    // For non-pure-Unix and non-UNC paths (this includes Windows absolute paths and relative paths):
    // 1. Ensure backslashes are used as separators.
    normalized = normalized.replace(/\//g, '\\');
    // 2. Ensure no internal double backslashes remain.
    //    The pre-normalize step for non-UNC should have handled most of this for `tempPath`.
    //    This ensures consistency after path.normalize might have reintroduced or missed some.
    normalized = normalized.replace(/\\\\/g, '\\');
  }

  // Capitalize drive letter if it's a Windows path starting with a lowercase drive letter.
  if (normalized.match(/^[a-z]:\\/)) {
    normalized = normalized.charAt(0).toUpperCase() + normalized.slice(1);
  }

  return normalized;
}
```

Comment (cliffhall) [reply src/filesystem/path-utils.ts]: With this refactored `normalizePath` function, I get 100% coverage of `path-utils.ts` and all tests passing on Darwin. I don't know about Windows. 

Screenshot 2025-06-19 at 5 38 37 PM

**Comment by olaservo** [comment]
> Those changes were a bit different from the suggested ones that passed locally.
> 
> I'm getting these errors now:

I made some updates and ran them on both Mac OS and Windows this time:

Screenshot 2025-06-19 at 9 31 45 PM

image

**Comment by olaservo** [comment]
Added checks for Typescript servers to run tests if found, here is an example run: https://github.com/modelcontextprotocol/servers/actions/runs/15780741769/job/44485530850?pr=1405

**Comment by cliffhall** [comment]
Screenshot 2025-06-20 at 3 50 45 PM

Tests ran, but we lost coverage. Did those previously suggested changes fail on Windows?

File: src/filesystem/package.json [hunk 2]
```diff
@@ -16,20 +16,26 @@
   "scripts": {
     "build": "tsc && shx chmod +x dist/*.js",
     "prepare": "npm run build",
-    "watch": "tsc --watch"
+    "watch": "tsc --watch",
+    "test": "jest --config=jest.config.cjs"
```
Comment (cliffhall) [inline src/filesystem/package.json]:
```suggestion
    "test": "jest --config=jest.config.cjs --coverage"
```

Comment (cliffhall) [reply src/filesystem/package.json]: Adding flag to get coverage report at end of test run.

**Review by cliffhall — CHANGES_REQUESTED** [review]

File: src/filesystem/path-utils.ts (hunk 1 above)
Comment (cliffhall) [reply src/filesystem/path-utils.ts]: BTW, didn't mean for it to fully include this request again. Will leave it for reference. Was just adding the flag below...

**Review by cliffhall — APPROVED** [review]
Approving to get a big win in for the day, but would be good to get coverage on line 76 of the path-utils.ts at some point. This is a massive improvement in confidence for the filesystem server.
